src/concrete_solve.jl

```diff
@@ -1342,9 +1416,18 @@ function DiffEqBase._concrete_solve_adjoint(
             end
         else
             # use TrackedArray for efficiency of the tape
-            _f(args...) = ArrayInterface.aos_to_soa(prob.f(args...))
+            _f(args...) = begin
+                res = prob.f(args...)
+                # reduce(vcat, vector_of_length_1) returns a Real
+                # this preserves the type of the returned vector
+                if length(res) == 1
+                    res
+                else
+                    reduce(vcat, res)
+                end
```

Comment:
How is this related to this PR? Also, doesn't ArrayInterface's function do this? If it doesn't, can't we fix it there?